src_lockfree/mpsc_queue.ml

```diff
@@ -1,152 +1,145 @@
+  | T Topen -> head
+  | T Tclosed -> assert false
+  | T (Snoc (xs, x)) -> rev_to (Cons (x, H head)) xs
+
+let rec rev (head) = function
+  | T Topen -> head
+  | T Tclosed -> assert false
+  | T (Snoc (xs, x)) -> rev (H (Cons (x, head))) xs
+
+let rev_pop = function
+  | (Snoc (tail, x) : (_, [< `Snoc ]) tail) -> rev_to (Cons (x, H Hopen)) tail
+
+let rec append a b =
+  match b with
+  | H Hclosed ->
+      failwith
+        "This cannot happen, maybe you are running close concurrently with \
+         another [close]"
+  | H Hopen -> a
+  | H (Cons (x, xs)) -> H (Cons (x, append a xs))
 
-let pop t =
-  let p = t.head in
-  (* [p] is the previously-popped item. *)
-  let node = Atomic.get p.next in
-  Node.fold node
-    ~none:(fun () -> raise Empty)
-    ~some:(fun node ->
-      t.head <- node;
-      let v = node.value in
-      node.value <- Obj.magic ();
-      (* So it can be GC'd *)
-      v)
+let pop_opt t =
+  match t.head with
+  | H Hclosed -> raise Closed (*Closed and empty*)
+  | H (Cons (x, xs)) ->
+      t.head <- xs;
+      Some x
+  | H Hopen -> (
+      match Atomic.exchange t.tail (T Topen) with
+      | T Topen -> None
+      | T Tclosed ->
+          failwith
+            "This cannot happen, maybe you are running close concurrently with \
+             [pop]"
+      | T (Snoc _ as tail) -> (
+          match rev_pop tail with
+          | Cons (x, xs) ->
+              t.head <- xs;
+              Some x))
 
-let peek_opt t =
-  let p = t.head in
-  (* [p] is the previously-popped item. *)
-  let node = Atomic.get p.next in
-  Node.fold node ~none:(fun () -> None) ~some:(fun node -> Some node.value)
+let pop t =
+  match t.head with
+  | H Hclosed -> raise Closed (*Closed and empty*)
+  | H (Cons (x, xs)) ->
+      t.head <- xs;
+      x
+  | H Hopen -> (
+      match Atomic.exchange t.tail (T Topen) with
+      | T Topen -> raise Empty
+      | T Tclosed ->
+          failwith
+            "This cannot happen, maybe you are running close concurrently with \
+             [pop]"
+      | T (Snoc _ as tail) -> (
+          match rev_pop tail with
+          | Cons (x, xs) ->
+              t.head <- xs;
+              x))
+
+let close t =
+  match Atomic.exchange t.tail (T Tclosed) with
+  | T Tclosed -> raise Closed
+  | tail ->
+      let y = rev (H Hclosed) tail in
+      t.head <- append y t.head
 
 let peek t =
-  let p = t.head in
-  (* [p] is the previously-popped item. *)
-  let node = Atomic.get p.next in
-  Node.fold node ~none:(fun () -> raise Empty) ~some:(fun node -> node.value)
+  match t.head with
+  | H (Cons (x, _)) -> x
+  | H Hclosed -> raise Closed
+  | H Hopen -> (
+      match Atomic.get t.tail with
+      | T Topen -> raise Empty
+      | T Tclosed ->
+          failwith
+            "This cannot happen, maybe you are running close concurrently with \
+             [close]"
+      | T (Snoc _ as tail) -> ( match rev_pop tail with (Cons (x, _)) -> x))
```

Comment:
This can be a potential performance issue, because repeated calls to `peek` keep on performing the `rev_pop` call.  IOW, it would make sense to move the elements from the tail to the head in peek so that peek will not repeatedly reverse the tail.